epoch_failover_test.go

```diff
@@ -765,6 +766,138 @@ func TestEpochLeaderFailoverNotNeeded(t *testing.T) {
 	require.False(t, timedOut.Load())
 }
 
+type rebroadcastComm struct {
+	nodes      []NodeID
+	emptyVotes chan *EmptyVote
+}
+
+func newRebroadcastComm(nodes []NodeID) *rebroadcastComm {
+	return &rebroadcastComm{
+		nodes:      nodes,
+		emptyVotes: make(chan *EmptyVote),
+	}
+}
+
+func (r *rebroadcastComm) ListNodes() []NodeID {
+	return r.nodes
+}
+
+func (r *rebroadcastComm) SendMessage(*Message, NodeID) {
+
+}
+
+func (r *rebroadcastComm) Broadcast(msg *Message) {
+	if msg.EmptyVoteMessage != nil {
+		r.emptyVotes <- msg.EmptyVoteMessage
+	}
+}
+
+func TestEpochRebroadcastsEmptyVote(t *testing.T) {
+	l := testutil.MakeLogger(t, 2)
+	bb := &testBlockBuilder{out: make(chan *testBlock, 1), blockShouldBeBuilt: make(chan struct{}, 1)}
+	storage := newInMemStorage()
+
+	nodes := []NodeID{{1}, {2}, {3}, {4}}
+
+	wal := newTestWAL(t)
+
+	epochTime := time.Now()
+	comm := newRebroadcastComm(nodes)
+	conf := EpochConfig{
+		MaxProposalWait:     DefaultMaxProposalWaitTime,
+		MaxRebrodcastWait:   DefaultEmptyVoteRebroadcastTimeout,
+		StartTime:           epochTime,
+		Logger:              l,
+		ID:                  nodes[3], // so we are not the leader
+		Signer:              &testSigner{},
+		WAL:                 wal,
+		Verifier:            &testVerifier{},
+		Storage:             storage,
+		Comm:                comm,
+		BlockBuilder:        bb,
+		SignatureAggregator: &testSignatureAggregator{},
+	}
+
+	e, err := NewEpoch(conf)
+	require.NoError(t, err)
+
+	require.NoError(t, e.Start())
+	require.Equal(t, uint64(0), e.Metadata().Round)
+	require.Equal(t, uint64(0), e.Metadata().Round)
+	require.False(t, wal.containsEmptyVote(0))
+
+	bb.blockShouldBeBuilt <- struct{}{}
+
+	// wait for the initial empty vote broadcast
+	done := false
+	for !done {
+		select {
+		case emptyVote := <-comm.emptyVotes:
+			require.True(t, wal.containsEmptyVote(0))
+			require.Equal(t, uint64(0), emptyVote.Vote.Round)
+			done = true
+		case <-time.After(10 * time.Millisecond):
+			epochTime = epochTime.Add(e.MaxRebrodcastWait)
+			e.AdvanceTime(epochTime)
+		}
+	}
+
+	// since the round does not advance, continue to rebroadcast
+	for range 10 {
+		epochTime = epochTime.Add(e.MaxRebrodcastWait)
+		e.AdvanceTime(epochTime)
+		emptyVote := <-comm.emptyVotes
+		require.Equal(t, uint64(0), emptyVote.Vote.Round)
+		wal.assertWALSize(1)
+	}
+
+	emptyNotarization := newEmptyNotarization(nodes, 0, 0)
+	e.HandleMessage(&simplex.Message{
+		EmptyNotarization: emptyNotarization,
+	}, nodes[2])
+
+	wal.assertNotarization(0)
+
+	// ensure the rebroadcast was canceled
+	epochTime = epochTime.Add(e.MaxRebrodcastWait * 2)
+	e.AdvanceTime(epochTime)
+
+	// ensure the timeout was canceled
+	require.Len(t, comm.emptyVotes, 0)
```

Comment:
since `emptyVotes` is non buffered, there can never be a time where its length is not zero. we therefore should initialize it to have at least a buffer of 1, in order for this check to be meaningful.